README
NAME
    Gentoo::Overlay::Group - A collection of Gentoo::Overlay objects.

VERSION
    version 1.000000

SYNOPSIS
    This is a wrapper around "Gentoo::Overlay" that makes it easier to
    perform actions on a group of overlays.

      my $group = Gentoo::Overlay::Group->new();
      $group->add_overlay('/usr/portage');
      $group->add_overlay('/usr/local/portage/');
      $group->iterate( packages => sub {
        my ( $self, $context ) = @_;
        # Traverse-Order:
        # ::gentoo
        #   category_a
        #     package_a
        #     package_b
        #   category_b
        #     package_a
        #     package_b
        # ::hentoo
        #   category_a
        #     package_a
        #     package_b
        #   category_b
        #     package_a
        #     package_b
      });

METHODS
  add_overlay
      $object->add_overlay( '/path/to/overlay' );
      $object->add_overlay( Path::Tiny::path( '/path/to/overlay' ) );
      $object->add_overlay( Gentoo::Overlay->new( path => '/path/to/overlay' ) );

  iterate
      $object->iterate( ebuilds => sub {


      });

ATTRIBUTE ACCESSORS
  overlay_names
      my @names = $object->overlay_names

  overlays
      my @overlays = $object->overlays;

  get_overlay
      my $overlay = $object->get_overlay('gentoo');

PRIVATE ATTRIBUTES
  _overlays
      isa => HashRef[ Gentoo__Overlay_Overlay ], ro, lazy

PRIVATE ATTRIBUTE ACCESSORS
  _has_overlay
      if( $object->_has_overlay('gentoo') ){
        Carp::croak('waah');
      }

  _set_overlay
      $object->_set_overlay( 'gentoo' => $overlay_object );

PRIVATE FUNCTIONS
  _type_print
    Lightweight flat dumper optimized for displaying user parameters in a
    format similar to a method signature.

      printf '[%s]', join q{,} , map { _type_print } @array

PRIVATE METHODS
  _iterate_ebuilds
      $object->_iterate_ebuilds( ignored => sub { } );

  _iterate_categories
      $object->_iterate_categories( ignored => sub { } );

  _iterate_packages
      $object->_iterate_packages( ignored => sub { } );

  _iterate_overlays
      $object->_iterate_overlays( ignored => sub { } );

  _add_overlay_object
      $groupobject->_add_overlay_object( $object );

  _add_overlay_gentoo_object
      $groupobject->_add_overlay_gentoo_object( $gentoo_object );

  _add_overlay_path_class
      $groupobject->_add_overlay_path_class( $path_class_object );

  _add_overlay_string_path
      $groupobject->_add_overlay_string_path( $path_string );

AUTHOR
    Kent Fredric <kentnl@cpan.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2014 by Kent Fredric <kentnl@cpan.org>.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Release history for Gentoo-Overlay-Group

{{$NEXT}}

1.000001 2017-03-08T11:49:04Z f4c64a9
 - Bugfix: Avoid test failures on -Ddefault_inc_excludes_dot

 [Dependencies::Stats]
 - Dependencies changed since 1.000000, see misc/*.deps* for details
 - configure: +1 (recommends: ↑1)
 - develop: +11 ↑3 -3 (recommends: +1, suggests: ↑2)
 - runtime: (recommends: +1)
 - test: +1 (recommends: +1 ↑1 ↓1)

1.000000 2014-10-08T06:59:17Z 23ddf32
 [00 Major - Version Scheme Change]
 - Migrating from 0.x.y to 1.xxxyyy
 - This may affect hand-operated downstreams
 - https://questhub.io/realm/perl/quest/5388d3a9bbd0be2224000129
 - This is the primary reason for the 1.x on the box.

 [01 Major - Installer Toolkit Change]
 - Migrating from Module::Build to ExtUtils::MakeMaker
 - This may affect hand-operated downstreams
 - Critical dependencies such as Test::More have "latest" dependencies migrated to recommendations only.
 - https://questhub.io/realm/perl/quest/5388cc12bbd0be5b2500011f

 [02 Major - Implemented in Moo]
 - Internals refactored and implemented in a Moo based stack instead of a Moose one.

 [Dependencies::Stats]
 - Dependencies changed since 0.2.0, see misc/*.deps* for details
 - build: -1
 - configure: +1 -1 (recommends: +1)
 - develop: +49 -2 (recommends: -1, suggests: +1 ↑1)
 - runtime: +5 ↑1 ↓1 -4
 - test: +2 ↓1 -2 (recommends: +3)

 [Metadata]
 - Migrate to kentnl org

0.2.0 2013-03-13T07:25:37Z
 [!Notable Changes]
 - Uses Path::Tiny instead of Path::Class
 - BACKCOMPAT: No longer accepts user constructed Path::Class objects as params

 [Dependencies::Stats]
 - Dependencies changed since 0.1.0, see misc/*.deps* for details
 - build: ↑1
 - configure: ↑1
 - develop: +5 (suggests: ↑1)
 - runtime: +1 ↑1 -1
 - test: +1 ↑1 -1

 [Documentation]
 - Copyright year
 - Add README.mkdn

 [Meta]
 - bugtracker to github issues

0.1.0 2012-06-21T13:32:12Z
 - First version.


Build.PL

use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "File::Find" => 0,
    "File::Temp" => 0,
    "FindBin" => 0,
    "Module::Build" => "0.4003",
    "Path::Tiny" => 0,
    "Test::Fatal" => 0,
    "Test::More" => "0.98",
    "Test::Output" => 0
  },
  "configure_requires" => {
    "Module::Build" => "0.4003"
  },
  "dist_abstract" => "A collection of Gentoo::Overlay objects.",
  "dist_author" => [
    "Kent Fredric <kentnl\@cpan.org>"
  ],
  "dist_name" => "Gentoo-Overlay-Group",
  "dist_version" => "0.2.0",
  "license" => "perl",
  "module_name" => "Gentoo::Overlay::Group",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "Gentoo::Overlay" => "v1.0.3",
    "Gentoo::Overlay::Exceptions" => 0,
    "Gentoo::Overlay::Types" => 0,
    "Moose" => 0,
    "MooseX::Has::Sugar" => 0,
    "MooseX::Types::Moose" => 0,
    "MooseX::Types::Path::Tiny" => 0,
    "Scalar::Util" => 0,
    "namespace::autoclean" => 0,
    "perl" => "v5.10.0",
    "strict" => 0,
    "warnings" => 0
  },
  "script_files" => []
);


my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


lib/Gentoo/Overlay/Group.pm
use 5.006;
use strict;
use warnings;

package Gentoo::Overlay::Group;

our $VERSION = '1.000002';

# ABSTRACT: A collection of Gentoo::Overlay objects.

# AUTHORITY

use Moo qw( has );
use MooX::HandlesVia;
use MooseX::Has::Sugar qw( ro lazy );
use Types::Standard qw( HashRef Str );
use Types::Path::Tiny qw( Dir );
use namespace::clean;

use Gentoo::Overlay 2.001001;
use Gentoo::Overlay::Types qw( Gentoo__Overlay_Overlay );
use Gentoo::Overlay::Exceptions qw( exception );
use Scalar::Util qw( blessed );

=p_attr _overlays

  isa => HashRef[ Gentoo__Overlay_Overlay ], ro, lazy

=p_attr_acc _has_overlay

  if( $object->_has_overlay('gentoo') ){
    Carp::croak('waah');
  }

=p_attr_acc _set_overlay

  $object->_set_overlay( 'gentoo' => $overlay_object );

=attr_acc overlay_names

  my @names = $object->overlay_names

=attr_acc overlays

  my @overlays = $object->overlays;

=attr_acc get_overlay

  my $overlay = $object->get_overlay('gentoo');

=cut

has '_overlays' => (
  ro, lazy,
  isa => HashRef [Gentoo__Overlay_Overlay],
  default     => sub { return {} },
  handles_via => 'Hash',
  handles     => {
    _has_overlay  => exists   =>,
    overlay_names => keys     =>,
    overlays      => elements =>,
    get_overlay   => get      =>,
    _set_overlay  => set      =>,
  },
);

my $_str = Str();

=p_func _type_print

Lightweight flat dumper optimized for displaying user parameters in a format similar to a method signature.

  printf '[%s]', join q{,} , map { _type_print } @array

=cut

sub _type_print {
  return
      ref $_     ? ref $_
    : defined $_ ? 'scalar<' . $_ . '>'
    :              'scalar=undef';

}

=method add_overlay

  $object->add_overlay( '/path/to/overlay' );
  $object->add_overlay( Path::Tiny::path( '/path/to/overlay' ) );
  $object->add_overlay( Gentoo::Overlay->new( path => '/path/to/overlay' ) );

=cut

sub add_overlay {
  my ( $self, @args ) = @_;
  if ( 1 == @args and blessed $args[0] ) {
    goto $self->can('_add_overlay_object');
  }
  if ( $_str->check( $args[0] ) ) {
    goto $self->can('_add_overlay_string_path');
  }
  return exception(
    ident   => 'bad overlay type',
    message => <<'EOF',
Unrecognised parameter types passed to add_overlay.
  Expected: \n%{signatures}s.
  Got: [%{type}s]}.
EOF
    payload => {
      signatures => ( join q{},  map { qq{    \$group->add_overlay( $_ );\n} } qw( Str Path::Tiny Gentoo::Overlay ) ),
      type       => ( join q{,}, map { _type_print } @args ),
    },
  );
}

=method iterate

  $object->iterate( ebuilds => sub {


  });

=cut

sub iterate {
  my ( $self, $what, $callback ) = @_;    ## no critic (Variables::ProhibitUnusedVarsStricter)
  my %method_map = (
    ebuilds    => _iterate_ebuilds    =>,
    categories => _iterate_categories =>,
    packages   => _iterate_packages   =>,
    overlays   => _iterate_overlays   =>,
  );
  if ( exists $method_map{$what} ) {
    goto $self->can( $method_map{$what} );
  }
  return exception(
    ident   => 'bad iteration method',
    message => 'The iteration method %{what_method}s is not a known way to iterate.',
    payload => { what_method => $what, },
  );
}

=p_method _iterate_ebuilds

  $object->_iterate_ebuilds( ignored => sub { } );

=cut

sub _iterate_ebuilds {
  my ( $self, undef, $callback ) = @_;
  my $real_callback = sub {
    my (%package_config) = %{ $_[1] };
    my $inner_callback = sub {
      my (%ebuild_config) = %{ $_[1] };
      $self->$callback( { ( %package_config, %ebuild_config ) } );
    };
    $package_config{package}->_iterate_ebuilds( ebuilds => $inner_callback );
  };
  $self->_iterate_packages( packages => $real_callback );
  return;
}

=p_method _iterate_categories

  $object->_iterate_categories( ignored => sub { } );

=cut

# categories = { /overlays/categories

sub _iterate_categories {
  my ( $self, undef, $callback ) = @_;
  my $real_callback = sub {
    my (%overlay_config) = %{ $_[1] };
    my $inner_callback = sub {
      my (%category_config) = %{ $_[1] };
      $self->$callback( { ( %overlay_config, %category_config ) } );
    };
    $overlay_config{overlay}->_iterate_categories( categories => $inner_callback );
  };
  $self->_iterate_overlays( overlays => $real_callback );
  return;
}

=p_method _iterate_packages

  $object->_iterate_packages( ignored => sub { } );

=cut

sub _iterate_packages {
  my ( $self, undef, $callback ) = @_;
  my $real_callback = sub {
    my (%category_config) = %{ $_[1] };
    my $inner_callback = sub {
      my (%package_config) = %{ $_[1] };
      $self->$callback( { ( %category_config, %package_config ) } );
    };
    $category_config{category}->_iterate_packages( packages => $inner_callback );
  };
  $self->_iterate_categories( categories => $real_callback );
  return;
}

=p_method _iterate_overlays

  $object->_iterate_overlays( ignored => sub { } );

=cut

# overlays = { /overlays }
sub _iterate_overlays {
  my ( $self, undef, $callback ) = @_;
  my %overlays     = $self->overlays;
  my $num_overlays = scalar keys %overlays;
  my $last_overlay = $num_overlays - 1;
  my $offset       = 0;
  for my $overlay_name ( sort keys %overlays ) {
    local $_ = $overlays{$overlay_name};
    $self->$callback(
      {
        overlay_name => $overlay_name,
        overlay      => $overlays{$overlay_name},
        num_overlays => $num_overlays,
        last_overlay => $last_overlay,
        overlay_num  => $offset,
      }
    );
    $offset++;
  }
  return;
}

my $_gentoo_overlay = Gentoo__Overlay_Overlay();
my $_path_class_dir = Dir();

# This would be better in M:M:TypeCoercion

=p_method _add_overlay_object

  $groupobject->_add_overlay_object( $object );

=cut

sub _add_overlay_object {
  my ( $self, $overlay, @rest ) = @_;

  if ( $_gentoo_overlay->check($overlay) ) {
    goto $self->can('_add_overlay_gentoo_object');
  }
  if ( $_path_class_dir->check($overlay) ) {
    goto $self->can('_add_overlay_path_class');
  }
  return exception(
    ident   => 'bad overlay object type',
    message => <<'EOF',
Unrecognised parameter object types passed to add_overlay.
  Expected: \n%{signatures}s.
  Got: [%{type}s]}.
EOF
    payload => {
      signatures => ( join q{}, map { qq{    \$group->add_overlay( $_ );\n} } qw( Str Path::Tiny Gentoo::Overlay ) ),
      type => ( join q{,}, blessed $overlay, map { _type_print } @rest ),
    },
  );
}

=p_method _add_overlay_gentoo_object

  $groupobject->_add_overlay_gentoo_object( $gentoo_object );

=cut

sub _add_overlay_gentoo_object {
  my ( $self, $overlay, ) = @_;
  $_gentoo_overlay->assert_valid($overlay);
  if ( $self->_has_overlay( $overlay->name ) ) {
    return exception(
      ident   => 'overlay exists',
      message => 'The overlay named %{overlay_name}s is already added to this group.',
      payload => { overlay_name => $overlay->name },
    );
  }
  $self->_set_overlay( $overlay->name, $overlay );
  return;
}

=p_method _add_overlay_path_class

  $groupobject->_add_overlay_path_class( $path_class_object );

=cut

sub _add_overlay_path_class {    ## no critic ( RequireArgUnpacking )
  my ( $self, $path, ) = @_;
  $_path_class_dir->assert_valid($path);
  my $go = Gentoo::Overlay->new( path => $path, );
  @_ = ( $self, $go );
  goto $self->can('_add_overlay_gentoo_object');
}

=p_method _add_overlay_string_path

  $groupobject->_add_overlay_string_path( $path_string );

=cut

sub _add_overlay_string_path {    ## no critic ( RequireArgUnpacking )
  my ( $self, $path_str, ) = @_;
  $_str->assert_valid($path_str);
  my $path = $_path_class_dir->coerce($path_str);
  @_ = ( $self, $path );
  goto $self->can('_add_overlay_path_class');
}

1;

=head1 SYNOPSIS

This is a wrapper around L<< C<Gentoo::Overlay>|Gentoo::Overlay >> that makes it easier to perform actions on a group of overlays.

  my $group = Gentoo::Overlay::Group->new();
  $group->add_overlay('/usr/portage');
  $group->add_overlay('/usr/local/portage/');
  $group->iterate( packages => sub {
    my ( $self, $context ) = @_;
    # Traverse-Order:
    # ::gentoo
    #   category_a
    #     package_a
    #     package_b
    #   category_b
    #     package_a
    #     package_b
    # ::hentoo
    #   category_a
    #     package_a
    #     package_b
    #   category_b
    #     package_a
    #     package_b
  });

=cut


Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.020.
use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker ;



my %WriteMakefileArgs = (
  "ABSTRACT" => "A collection of Gentoo::Overlay objects.",
  "AUTHOR" => "Kent Fredric <kentnl\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Gentoo-Overlay-Group",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "Gentoo::Overlay::Group",
  "PREREQ_PM" => {
    "Gentoo::Overlay" => "2.001001",
    "Gentoo::Overlay::Exceptions" => 0,
    "Gentoo::Overlay::Types" => 0,
    "Moo" => 0,
    "MooX::HandlesVia" => 0,
    "MooseX::Has::Sugar" => 0,
    "Scalar::Util" => 0,
    "Types::Path::Tiny" => 0,
    "Types::Standard" => 0,
    "namespace::clean" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "FindBin" => 0,
    "Path::Tiny" => 0,
    "Test::Fatal" => 0,
    "Test::More" => "0.96",
    "Test::Output" => 0
  },
  "VERSION" => "1.000000",
  "test" => {
    "TESTS" => "t/*.t t/00-compile/*.t"
  }
);


my %FallbackPrereqs = (
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "FindBin" => 0,
  "Gentoo::Overlay" => "2.001001",
  "Gentoo::Overlay::Exceptions" => 0,
  "Gentoo::Overlay::Types" => 0,
  "Moo" => 0,
  "MooX::HandlesVia" => 0,
  "MooseX::Has::Sugar" => 0,
  "Path::Tiny" => 0,
  "Scalar::Util" => 0,
  "Test::Fatal" => 0,
  "Test::More" => "0.96",
  "Test::Output" => 0,
  "Types::Path::Tiny" => 0,
  "Types::Standard" => 0,
  "namespace::clean" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



